src/sources/file/line_agg.rs

```diff
@@ -0,0 +1,519 @@
+    /// This is useful where a log line contains a marker indicating that it
+    /// begins a new message.
+    HaltBefore,
+
+    /// All consecutive lines, up to and including the first line matching this
+    /// pattern, are included in the group.
+    /// This is useful where a log line ends with a termination marker, such as
+    /// a semicolon.
+    HaltWith,
+}
+
+#[derive(Debug, Clone)]
+pub(super) struct Config {
+    /// Start pattern to look for as a beginning of the message.
+    pub start_pattern: Regex,
+    /// Condition pattern to look for. Exact behavior is configured via `mode`.
+    pub condition_pattern: Regex,
+    /// Mode of operation, specifies how the condition pattern is interpreted.
+    pub mode: Mode,
+    /// The maximum time to wait for the continuation. Once this timeout is
+    /// reached, the buffered message is guaraneed to be flushed, even if
+    /// incomplete.
+    pub timeout: Duration,
+}
+
+impl Config {
+    pub(super) fn for_legacy(marker: Regex, timeout_ms: u64) -> Self {
+        let start_pattern = marker;
+        let condition_pattern = start_pattern.clone();
+        let mode = Mode::HaltBefore;
+        let timeout = Duration::from_millis(timeout_ms);
+
+        Self {
+            start_pattern,
+            condition_pattern,
+            mode,
+            timeout,
+        }
+    }
+}
+
+/// Private type alias to be more expressive in the internal implementation.
+type Filename = String;
+
+pub(super) struct LineAgg<T> {
+    /// The stream from which we read the lines.
+    inner: T,
+
+    /// Configuration parameters to use.
+    config: Config,
+
+    /// Line per filename.
+    buffers: HashMap<Filename, BytesMut>,
+
+    /// Draining queue. We switch to draining mode when we get `None` from
+    /// the inner stream. In this mode we stop polling `inner` for new lines
+    /// and just flush all the buffered data.
+    draining: Option<Vec<(Bytes, Filename)>>,
+
+    /// A queue of filename timeouts.
+    timeouts: DelayQueue<Filename>,
+
+    /// A queue of filenames with expired timeouts.
+    expired: VecDeque<Filename>,
+}
+
+impl<T> LineAgg<T> {
+    pub(super) fn new(inner: T, config: Config) -> Self {
+        Self {
+            inner,
+
+            config,
+
+            draining: None,
+            buffers: HashMap::new(),
+            timeouts: DelayQueue::new(),
+            expired: VecDeque::new(),
+        }
+    }
+}
+
+impl<T> Stream for LineAgg<T>
+where
+    T: Stream<Item = (Bytes, Filename), Error = ()>,
+{
+    /// `Bytes` - the message data; `Filename` - file name.
+    type Item = (Bytes, Filename);
+    type Error = ();
+
+    fn poll(&mut self) -> Poll<Option<Self::Item>, Self::Error> {
+        loop {
+            // If we're in draining mode, short circut here.
+            if let Some(to_drain) = &mut self.draining {
+                if let Some((line, src)) = to_drain.pop() {
+                    return Ok(Async::Ready(Some((line, src))));
```

Comment:
Since this function is just one whole `loop`, I wonder if it wouldn't be better to `break VALUE` instead of `return VALUE` for this and all return statements within. That's a bit of a style thing though, so I don't have a strong opinion on this.

Reply:
Is there a reason why you prefer `break` over `return` in such cases? I understand, in this particular case it doesn’t make any practical difference at all, but there may be reasons - like `break`s being better in typical ways of how such loops evolve.

Reply:
I have tended to prefer the expression or functional notation over procedural. I agree this makes no practical difference here.